### PR TITLE
Test/components/losses

### PR DESCRIPTION
**Pull request by siddarthsingh1**
## What?
Unit testing for losses component
## Why?
Increasing test coverage for Mava redesign
## Extra
~~Testing for the consistency of the output of the loss is not completed.~~

### CODE REVIEW

**Review by mmorris44 — CHANGES_REQUESTED** [review]
Thanks @siddarthsingh1! Please see my comments.

File: tests/jax/components/training/losses_test.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
+from types import SimpleNamespace
+
+# from typing import Callable, List, Tuple
```
[the remark this answers is not in the document]
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py:3]: Updated

File: tests/jax/components/training/losses_test.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+import pytest
+
+from mava.components.jax.training.losses import (
+    MAPGTrustRegionClippingLossConfig,
+    MAPGWithTrustRegionClippingLoss,
+)
+from mava.systems.jax.trainer import Trainer
+
+
+class mock_net:
```
[the remark this answers is not in the document]
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py:17]: Updated

File: tests/jax/components/training/losses_test.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+) -> None:
+    """Test whether mapg loss output is as expected"""
+    mapg_trust_region_clipping_loss.on_training_loss_fns(trainer=mock_trainer)
+    grad_fn = mock_trainer.store.grad_fn
+
+    actions = mock_actions()
+    behaviour_log_probs = mock_behaviour_log_probs()
+    target_values = mock_target_values()
+    advantages = mock_advantages()
+    behavior_values = mock_behavior_values()
```
[the remark this answers is not in the document]
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py:189]: Updated

File: tests/jax/components/training/losses_test.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+) -> None:
+    """Test whether mapg function is successfully created"""
+
+    mapg_trust_region_clipping_loss.on_training_loss_fns(trainer=mock_trainer)
+    assert hasattr(mock_trainer.store, "grad_fn")
+    assert isinstance(mock_trainer.store.grad_fn, Callable)  # type:ignore
+
+
+def test_mapg_config_creation() -> None:
+    """Test whether mapg loss config variables are of correct type"""
```
[the remark this answers is not in the document]
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py:167]: Removed

File: tests/jax/components/training/losses_test.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+def entropy(distribution_params: jnp.array) -> jnp.ndarray:
+    """Mock function to return fixed entropy"""
+
+    entropy = jnp.array([-1.7, -1.7, -1.7, -1.7, -1.7])
+
+    return entropy
+
+
+def mock_actions() -> Dict[str, jnp.ndarray]:
```
[the remark this answers is not in the document]
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py:45]: Updated

File: tests/jax/components/training/losses_test.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+            [-1.7, -1.7, -1.7, -1.7, -1.7]
+        )
+
+
+def log_prob(distribution_params: jnp.array, actions: jnp.array) -> jnp.ndarray:
+    """Mock function to return fixed log probs"""
+
+    log_probs = jnp.array([-1.7, -1.7, -1.7, -1.7, -1.7])
+
+    return log_probs
```
[the remark this answers is not in the document]
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py:34]: Updated

File: tests/jax/components/training/losses_test.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+class mock_net:
+    """Creates a mock network for loss function"""
+
+    def apply(
+        observation: jnp.array, rng_key: jnp.array
+    ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+        """Mock function to apply the network to training data"""
+        return jnp.array([-1.7, -1.7, -1.7, -1.7, -1.7]), jnp.array(
+            [-1.7, -1.7, -1.7, -1.7, -1.7]
```
[the remark this answers is not in the document]
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py]: It's changed to use the variables to create the return.

File: tests/jax/components/training/losses_test.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    agent_0_loss = loss_info["agent_0"]
+    loss_entropy = agent_0_loss["loss_entropy"]
+    loss_policy = agent_0_loss["loss_policy"]
+    loss_total = agent_0_loss["loss_total"]
+    loss_value = agent_0_loss["loss_value"]
+
+    assert loss_entropy == 1.7
+    assert loss_policy == -2.0
+    assert loss_total == 9.062
+    assert loss_value == 22.09
```
[the remark this answers is not in the document]
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py]: For the rlax reliant values it checks if a different set of variables returns a lower value or not.

File: tests/jax/components/training/losses_test.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+import pytest
+
+from mava.components.jax.training.losses import (
+    MAPGTrustRegionClippingLossConfig,
+    MAPGWithTrustRegionClippingLoss,
+)
+from mava.systems.jax.trainer import Trainer
+
+
+class Mock_net:
```
Comment (mmorris44) [inline tests/jax/components/training/losses_test.py:15]: `MockNet`

File: tests/jax/components/training/losses_test.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+    def apply(
+        parameters: jnp.array, observation: jnp.array
+    ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+        """Mock function to apply the network to training data"""
+        return jnp.array(observation), jnp.array(observation)
+
+
+def log_prob(distribution_params: jnp.array, actions: jnp.array) -> jnp.ndarray:
+    """Mock function to return fixed log probs"""
+    return distribution_params + actions
```
Comment (mmorris44) [inline tests/jax/components/training/losses_test.py:27]: Can we not just concatenate here as well?

File: tests/jax/components/training/losses_test.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+
+class Mock_net:
+    """Creates a mock network for loss function"""
+
+    def apply(
+        parameters: jnp.array, observation: jnp.array
+    ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+        """Mock function to apply the network to training data"""
+        return jnp.array(observation), jnp.array(observation)
```
Comment (mmorris44) [inline tests/jax/components/training/losses_test.py:22]: They're already jnp arrays

File: tests/jax/components/training/losses_test.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+    low_agent_0_loss = low_loss_info["agent_0"]
+    low_loss_value = low_agent_0_loss["loss_value"]
+    low_loss_policy = low_agent_0_loss["loss_policy"]
+
+    assert loss_entropy == -0.47500002
+    assert loss_value == 6.4075003
+    assert loss_total == (loss_entropy * 0.01 + loss_policy + loss_value * 0.5)
+
+    assert low_loss_value < loss_value
+    assert low_loss_policy < loss_policy
```
Comment (mmorris44) [inline tests/jax/components/training/losses_test.py:191]: This is still confusing me a little bit, since we're asserting for a value of `loss_total` here?

**Review by mmorris44 — CHANGES_REQUESTED** [review]
Thanks @siddarthsingh1 for the changes since last review. Just a couple more things to fix that I found.

File: tests/jax/components/training/losses_test.py (hunk 9 above)
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py]: Changed

File: tests/jax/components/training/losses_test.py (hunk 10 above)
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py:27]: Since they're both arrays I added them together to pretend we're using them to calculate the log_probs. If they're concatenated it would change the output size.

File: tests/jax/components/training/losses_test.py (hunk 11 above)
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py]: There was an error for `observation = SimpleNamespace(observation= jnp.array([0.5, 0.5, 0.7, 0.2]))`  where I didn't set the observation to be a jnp array. It's been fixed and the extra `jnp.array(...)`s have been removed.

File: tests/jax/components/training/losses_test.py (hunk 12 above)
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py]: I was checking if the `loss_total` was being added together correctly based on the default coefficients in the config. I've removed it if you think it's necessary?

Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py]: I've changed it to check if the `low_loss_total' is less than the `loss_total` value.

File: tests/jax/components/training/losses_test.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+
+
+class MockNet:
+    """Creates a mock network for loss function"""
+
+    def apply(
+        parameters: jnp.array, observation: jnp.array
+    ) -> Tuple[jnp.ndarray, jnp.ndarray]:
+        """Mock function to apply the network to training data"""
+        return observation, observation
```
Comment (mmorris44) [inline tests/jax/components/training/losses_test.py:22]: `return parameters, observation` surely?

**Review by mmorris44 — CHANGES_REQUESTED** [review]
One final thing that I caught, sorry. Also the unresolved comment on `rlax` / `loss_total`

File: tests/jax/components/training/losses_test.py (hunk 12 above)
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py]: It gets the `policy_loss` value from rlax but it's just a summation of the other 3 losses. So if the other 3 losses are working as intended then the check is just too see that it's being generated as intended.

File: tests/jax/components/training/losses_test.py (hunk 13 above)
Comment (mmorris44) [reply tests/jax/components/training/losses_test.py:22]: But can't you just assert for the value of the parameters in the output?

File: tests/jax/components/training/losses_test.py (hunk 12 above)
Comment (mmorris44) [reply tests/jax/components/training/losses_test.py]: @siddarthsingh1 this still requires attention. You can add back in the direct assertion for loss values.

Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py]: Added it back

File: tests/jax/components/training/losses_test.py (hunk 13 above)
Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py]: It's actually supposed to be a dictionary. It's been changed to use the parameters value.Thanks for catching that!

File: tests/jax/components/training/losses_test.py (hunk 12 above)
Comment (mmorris44) [reply tests/jax/components/training/losses_test.py]: Awesome, if we're doing this then, do we even still need the `low_loss_total` etc. assertions?

File: tests/jax/components/training/losses_test.py [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+        agent: jnp.array([1.0, 1.0, 1.0, 1.0])
+        for agent in {"agent_0", "agent_1", "agent_2"}
+    }
+
+    low_target_values = {
+        agent: jnp.array([1.0, 1.0, 1.0, 1.0])
+        for agent in {"agent_0", "agent_1", "agent_2"}
+    }
+    low_advantages = {
+        agent: jnp.array([3.0, 3.0, 3.0, 3.0])
```
Comment (RuanJohn) [inline tests/jax/components/training/losses_test.py:153]: Maybe set these to be all 1.0 to make them lower than `advantages`

Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py:153]: The naming is confusing I think? They're used to generate the lower losses so they need to be higher to create a lower `loss_total`.

File: tests/jax/components/training/losses_test.py (hunk 12 above)
Comment (mmorris44) [reply tests/jax/components/training/losses_test.py]: @siddarthsingh1 what are your thoughts? Seems to me that if we can directly assert for values like above, then we don't need this high / low loss value setup

Comment (mmorris44) [reply tests/jax/components/training/losses_test.py]: Thanks for clarifying this morning. I'm happy with how it is now, makes sense!

**Review by mmorris44 — DISMISSED** [review]
Looks good! Thanks @siddarthsingh1

File: tests/jax/components/training/losses_test.py (hunk 14 above)
Comment (RuanJohn) [reply tests/jax/components/training/losses_test.py:153]: I see. Thank you for clarifying.

File: tests/jax/components/training/losses_test.py [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+    observations = {
+        "agent_0": observation,
+        "agent_1": observation,
+        "agent_2": observation,
+    }
+
+    agent_net_keys = {
+        "agent_0": "network_agent_0",
+        "agent_1": "network_agent_1",
+        "agent_2": "network_agent_2",
```
Comment (RuanJohn) [inline tests/jax/components/training/losses_test.py:51]: Please rename the 'agent_net_keys' values here to all be `network_agent` to reflect the shared network defined below.

Comment (siddarthsingh1) [reply tests/jax/components/training/losses_test.py]: Updated

**Review by RuanJohn — APPROVED** [review]
Thanks @siddarthsingh1!

**Review by mmorris44 — APPROVED** [review]
